File: 2015-elixir/lib/solutions/day_02.ex
defmodule Aoc.Solution.Day02 do
  @name "I Was Told There Would Be No Math"
  @behaviour Solution

  @impl Solution
  def get_name, do: @name

  @impl Solution
  def parse!(raw) do
    raw
    |> String.split()
    |> Enum.map(fn p -> p |> String.split("x") |> Enum.map(&String.to_integer/1) end)
  end

  @impl Solution
  def solve_first_part(input) do
    input
    |> Enum.map(fn p -> area(p) end)
    |> Enum.sum()
  end

  @impl Solution
  def solve_second_part(input) do
    input
    |> Enum.map(fn p -> ribbon(p) end)
    |> Enum.sum()
  end

  defp area(p) do
    [l, w, h] = p
    extra = Enum.min([l * w, w * h, h * l])
    2 * l * w + 2 * w * h + 2 * h * l + extra
  end

  defp ribbon(p) do
    [l, w, h] = p
    [x, y | _] = Enum.sort(p)
    ribbon = 2 * x + 2 * y
    bow = l * w * h
    bow + ribbon
  end
end

File: 2015-python/solutions/day_02.py
from solutions import BaseSolution


class Solution(BaseSolution):
    input_file = "02.txt"

    def __str__(self):
        return "Day 2: I Was Told There Would Be No Math"

    def solve(self, pi):
        o = self._solve(pi)
        return o[0]

    def solve_again(self, pi):
        o = self._solve(pi)
        return o[1]

    def _solve(self, pi):
        p1 = 0
        p2 = 0
        for p in pi.splitlines():
            l, w, h = sorted([int(s) for s in p.split("x")])
            p1 += 2 * (l * w + w * h + h * l) + min([l * w, w * h, h * l])
            p2 += 2 * l + 2 * w + l * w * h
        return p1, p2

    def parse_input(self, data):
        return data.strip()


if __name__ == "__main__":
    solution = Solution()
    solution.show_results()
